### PR TITLE
Trang Frego - Hotel

### PR DESCRIPTION
**Pull request by tfrego**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to decide if I wanted to store my booked dates within the Room class or Reservation class. My final decision was based on the fact that I needed to make a BlockOfRooms class that inherited from Reservation. With dates required for BlockOfRooms, I decided to store the dates in the Reservation class.
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained more concept on using inheritance and how it inherits all of the instance variables from the parent class.
Describe a _nominal test_ that you wrote for this assignment. | A nominal test consists of a reservation request for a room that was available.
Describe an _edge case test_ that you wrote for this assignment. | An edge case test consisted of a reservation request when the hotel is fully booked.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | When I wasn't correcting my program or going back and changing my design, the approach of writing pseudocode first, tests, then code worked great. When I had to change things, it was hard to follow those steps.

### CODE REVIEW

File: lib/booking_system.rb [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      end
+    end
+    if new_reservation != nil
+      return new_reservation
+    else
+      raise ArgumentError, 'no rooms available in date range'
+    end
+  end
+
+  def date_range(check_in_date, check_out_date)
```
Comment (CheezItMan) [inline lib/booking_system.rb:70]: I might make this method private, as it's a good helper.

File: lib/booking_system.rb [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  def assign_res_id
+    return @reservations.count + 1
+  end
+
+  # access the list of reservations for a specific date
+  def reservations_by_date(date)
+    date = Date.parse(date)
+    list_res = []
+    @reservations.each do |reservation|
+      if reservation.check_in_date <= date && reservation.check_out_date >= date
```
Comment (CheezItMan) [inline lib/booking_system.rb:91]: I think that you should have had:
`reservation.check_out_date >= date` changed to `reservation.check_out_date > date`

Because you can check in on the same date someone else checks out.

File: spec/booking_system_spec.rb [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+require 'pry'
+
+describe 'BookingSystem class' do
+  before do
+    @booking = BookingSystem.new
+    @booking.reserve_room("Oct 4 2018", "Oct 7 2018")
+    @booking.create_block_of_rooms("Dec 1 2018", "Dec 5 2018", 180)
+  end
+
+  let (:overlap_before) {
```
Comment (CheezItMan) [inline spec/booking_system_spec.rb:11]: nice clear set of `let` blocks with good variable names to use in testing.

File: spec/booking_system_spec.rb [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  describe 'load_rooms' do
+    it 'is an instance of Room' do
+      expect(@booking.rooms.first).must_be_kind_of Room
+    end
+  end
+
+  describe 'list_all_rooms' do
+    it 'returns a list of all the rooms in the hotel' do
+      expect(@booking.list_all_rooms).must_be_kind_of Array
+      expect(@booking.list_all_rooms[0]).must_be_kind_of Room
```
Comment (CheezItMan) [inline spec/booking_system_spec.rb]: You could also test the right number of rooms are loaded.

File: spec/booking_system_spec.rb [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+      expect(room1_after).must_equal room1_before
+      expect(room2_after).must_equal room2_before
+      expect(room3_after).must_equal room3_before
+      expect(room4_after).must_equal room4_before
+      expect(room5_after).must_equal room5_before
+      expect(room6_after).must_equal room6_before + 1
+    end
+  end
+
+  describe 'date_range' do
```
Comment (CheezItMan) [inline spec/booking_system_spec.rb:235]: Since the user would never be calling this method (it's really a method used internally to the booking system, I would make the method `private` and not test it directly.

File: spec/booking_system_spec.rb [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+      start_date = "Oct 4 2018"
+      end_date = "Oct 7 2018"
+      discounted_rate = 180
+      my_block = @booking.create_block_of_rooms(start_date, end_date, 180)
+
+      expect(my_block).must_be_kind_of BlockOfRooms
+      expect(my_block.check_in_date).must_equal Date.parse(start_date)
+      expect(my_block.check_out_date).must_equal Date.parse(end_date)
+      expect(my_block.room_cost).must_equal discounted_rate
+      expect(my_block.collection_rooms.count).must_equal 5
```
Comment (CheezItMan) [inline spec/booking_system_spec.rb:316]: If it's equal to 5, then it's going to be <= 5.

File: lib/reservation.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    return duration
+  end
+
+  def total_cost
+    total = length_of_stay * @room_cost
+    return total
+  end
+end
+
+class BlockOfRooms < Reservation
```
Comment (CheezItMan) [inline lib/reservation.rb]: This really should be in it's own file.

File: lib/reservation.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  attr_accessor :collection_rooms
+
+  def initialize(check_in_date, check_out_date, id: 0, room_cost: 200)
+    @check_in_date = Date.parse(check_in_date)
```
Comment (CheezItMan) [inline lib/reservation.rb]: wouldn't a call to `super` be appropriate?

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commits and good commit messages
Answer comprehension questions	|  Check, I understand it's hard to do pseudocode when refactoring
**Design** |
Each class is responsible for a single piece of the program |  Mostly, you (like I'm noting below) have both rooms and `BookingSystem` tracking the reservations.  
Classes are loosely coupled |  You do have both `BookingSystem` and `Room` tracking lists of reservations.  This can make the project awkward to change later.  
**Wave 1** |
List rooms |  Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error |  Check
**Wave 2** |
View available rooms for a given date range |  Check, with `unreserved_rooms_by_date`
Reserving a room that is not available produces an error | Check, but I wouldn't raise an `ArgumentError` instead I would raise a custom error or if I don't have that `StandardError`
**Wave 3** |
Create a block of rooms |  Check
Check if a block has rooms |  Check
Reserve a room from a block |  Check
Test coverage | 99.6%
**Fundamentals** | 
Names variables, classes and modules appropriately  | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Very well done
Appropriately uses methods to break down tasks into smaller simpler tasks | Check
Understands the differences between class and instance methods| Check
Appropriately uses iterators and enumerables| Check
Appropriately writes and utilizes classes| Check
Appropriately utilizes modules as a mechanism to encapsulate functionality| Check
**Wrap Up** |
There is a refactors.txt file | Check, I'm not clear on what all your refactors mean.  
The file provides a roadmap to future changes |  Some ideas for refactoring, but I don't understand what it all mentions.  
**Additional Feedback** |  Very well done, neat tricks with ranges, and pretty good organization.  Question:  Did you really gain much by making `BlockOfRooms` inherit from `Reservation`?  Did that simplify anything?  See my inline notes, but very nicely done.  You wrote excellent tests btw.
